fiaas_deploy_daemon/deployer/kubernetes/ingress.py

```diff
@@ -48,18 +51,62 @@ def delete(self, app_spec):
         LOG.info("Deleting ingresses for %s", app_spec.name)
         self._ingress_adapter.delete_list(app_spec)
 
+    def _map_hostnames_to_certificate_secrets(self, app_spec):
+        k8s_ingresses = self._ingress_adapter.find(app_spec)
+
+        hosts_map = {}
+        for k8s_ingress in k8s_ingresses:
+            for tls_obj in k8s_ingress.spec.tls:
+                if tls_obj.secretName:
+                    for host_item in tls_obj.hosts:
+                        hostname = host_item.strip()
+                        hosts_map[hostname] = tls_obj.secretName
+        return hosts_map
+
+    def _copy_secret(self, secret_name, new_name, app_spec):
+        try:
+            old_secret = Secret.get(secret_name, app_spec.namespace)
+        except NotFound:
+            old_secret = None
+        if old_secret:
+            new_metadata = ObjectMeta(
+                annotations=old_secret.metadata.annotations,
+                labels=old_secret.metadata.labels,
+                name=new_name,
+                namespace=old_secret.metadata.namespace
+            )
+            new_secret = Secret(metadata=new_metadata, data=old_secret.data, type=old_secret.type)
+            new_secret.save()
+
```

Comment:
finding which secret to copy and copying the secret seem to belong to the IngressTLSDeployer in terms of responsability instead of the main IngressDeployer. We will need to get all ingresses for each ingress we create, but that's a performance penalty worth considering in exchange for code clarity (e.g. one extra "find" call for each ingress, and the common case is to have just one ingress in the app). What are your thoughts on this?

Reply:
After a bit of thinking, we agreed to move the copy secrets code to the TLSDeployer, but keeping the hosts map function in place, to improve efficiency. WDYT?